bytes should translate to &[u8]

Per a discussion in the gitter chat, it seems that currently the proper traits are not implemented to allow for Python's `bytes` objects to be automatically translated into `&[u8]` parameters. Right now, in order to write a function that takes `bytes`, you need to use `&PyBytes`:

```rust
#[pyfunction]
fn print_bytes(_py: Python, x: &PyBytes) {
    println!("{:?}", x.as_bytes())
}
```

If you instead use `&[u8]` directly, there's an error at compile time because the `pyo3::PyTypeInfo` trait is not implemented for `[u8]`. I think it should be possible to pass both a `bytes` and `bytearray` object into a Rust function that takes `&[u8]`.

I am using version 0.6.0-alpha.4, and with this code:

```rust
#[pyfunction]
fn print_bytes(_py: Python, x: &[u8]) {
    println!("{:?}", x.as_bytes())
}
```

I get this error message:

```
error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
  --> src/lib.rs:36:1
   |
36 | #[pyfunction]
   | ^^^^^^^^^^^^^ doesn't have a size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `[u8]`
   = note: to learn more, visit <...>
   = note: required because of the requirements on the impl of
           `pyo3::FromPyObject<'_>` for `&[u8]`

error[E0277]: the trait bound `[u8]: pyo3::PyTypeInfo` is not satisfied
  --> src/lib.rs:36:1
   |
36 | #[pyfunction]
   | ^^^^^^^^^^^^^ the trait `pyo3::PyTypeInfo` is not implemented for `[u8]`
   |
   = note: required because of the requirements on the impl
           of `pyo3::PyTryFrom<'_>` for `[u8]`
   = note: required because of the requirements on the impl
           of `pyo3::FromPyObject<'_>` for `&[u8]`

error: aborting due to 2 previous errors
```

CC @thedrow

--- a/src/types/string.rs
+++ b/src/types/string.rs
@@ -215,6 +215,18 @@ impl<'source> FromPyObject<'source> for String {
     }
 }
 
+impl<'a> FromPy<&'a [u8]> for PyObject {
+    fn from_py(other: &'a [u8], py: Python) -> Self {
+        PyBytes::new(py, other).to_object(py)
+    }
+}
+
+impl<'a> FromPyObject<'a> for &'a [u8] {
+    fn extract(obj: &'a PyAny) -> PyResult<Self> {
+        Ok(<PyBytes as PyTryFrom>::try_from(obj)?.as_bytes())
+    }
+}
+
 #[cfg(test)]
 mod test {
     use super::{PyBytes, PyString};
@@ -242,6 +254,16 @@ mod test {
 
         let s2: &str = FromPyObject::extract(py_string.as_ref(py).into()).unwrap();
         assert_eq!(s, s2);
+    }
+
+    #[test]
+    fn test_extract_bytes() {
+        let gil = Python::acquire_gil();
+        let py = gil.python();
+
+        let py_bytes = py.eval("b'Hello Python'", None, None).unwrap();
+        let bytes: &[u8] = FromPyObject::extract(py_bytes).unwrap();
+        assert_eq!(bytes, b"Hello Python");
     }
 
     #[test]

--- a/tests/test_bytes.rs
+++ b/tests/test_bytes.rs
@@ -1,0 +1,40 @@
+use pyo3::prelude::*;
+use pyo3::types::PyBytes;
+use pyo3::wrap_pyfunction;
+
+mod common;
+
+#[pyfunction]
+fn get_pybytes() -> PyObject {
+    let gil = Python::acquire_gil();
+    let py = gil.python();
+
+    PyBytes::new(py, b"Hello World").into()
+}
+
+#[pyfunction]
+fn bytes_pybytes_conversion(bytes: &[u8]) -> &[u8] {
+    bytes
+}
+
+#[test]
+fn test_get_pybytes() {
+    let gil = Python::acquire_gil();
+    let py = gil.python();
+
+    let get_bytes = wrap_pyfunction!(get_pybytes)(py);
+    py_assert!(py, get_bytes, "get_bytes() == b'Hello World'");
+}
+
+#[test]
+fn test_pybytes_bytes_conversion() {
+    let gil = Python::acquire_gil();
+    let py = gil.python();
+
+    let bytes_pybytes_conversion = wrap_pyfunction!(bytes_pybytes_conversion)(py);
+    py_assert!(
+        py,
+        bytes_pybytes_conversion,
+        "bytes_pybytes_conversion(b'Hello World') == b'Hello World'"
+    );
+}

--- a/tests/test_bytes.rs
+++ b/tests/test_bytes.rs
@@ -1,29 +1,11 @@
 use pyo3::prelude::*;
-use pyo3::types::PyBytes;
 use pyo3::wrap_pyfunction;
 
 mod common;
 
 #[pyfunction]
-fn get_pybytes() -> PyObject {
-    let gil = Python::acquire_gil();
-    let py = gil.python();
-
-    PyBytes::new(py, b"Hello World").into()
-}
-
-#[pyfunction]
 fn bytes_pybytes_conversion(bytes: &[u8]) -> &[u8] {
     bytes
-}
-
-#[test]
-fn test_get_pybytes() {
-    let gil = Python::acquire_gil();
-    let py = gil.python();
-
-    let get_bytes = wrap_pyfunction!(get_pybytes)(py);
-    py_assert!(py, get_bytes, "get_bytes() == b'Hello World'");
 }
 
 #[test]

